torch/testing/_internal/composite_compliance.py

```diff
@@ -0,0 +1,174 @@
+import torch
+import contextlib
+from typing import Iterator
+from torch.utils._pytree import tree_map
+from functools import partial
+from torch.utils._python_dispatch import enable_python_mode
+
+# TODO: move this into library proper
+@contextlib.contextmanager
+def no_dispatch() -> Iterator[None]:
+    guard = torch._C._DisableTorchDispatch()  # type: ignore[attr-defined]
+    try:
+        yield
+    finally:
+        del guard
+
+def check_metadata_consistency(wrapper_tensor):
+    if not isinstance(wrapper_tensor, CompositeCompliantTensor):
+        return
+    elem = wrapper_tensor.elem
+    if wrapper_tensor.shape != elem.shape:
+        raise RuntimeError(
+            "This operator is not CompositeImplicitAutograd compliant: the "
+            "shape of the tensor was modified directly without "
+            "going through the PyTorch dispatcher.")
+    if wrapper_tensor.dtype != elem.dtype:
```

Comment:
I have a nagging feeling that this list of metadata is incomplete, but it's probably easy enough to add to over time

Reply:
Yeah, it's definitely not complete and we can add more over time.

Reply:
Isn't this simply a matter of going over all the fields in TensorImpl? This is decidedly finite.

Reply:
Went over TensorImpl fields and added the ones that were exposed in Python, with some exception... (see other comment)